### PR TITLE
[WIP] Add BT based implementation of waypoint_follower 

### PR DESCRIPTION
**Pull request by ymd-stella**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | related to https://github.com/ros-planning/navigation2/issues/1651 (another implementation of #803) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | None |

---

## Description of contribution in a few bullet points

* Add BT based implementation of waypoint_follower (Action server of FollowWaypoints)
  * Customizable recovery behavior will be available for FollowWaypoints (As in the case of NavigateToPose)
    * [follow_waypoints_with_skip.xml](https://github.com/ymd-stella/navigation2/blob/020d6f8ff4dfb9d83da491b193cb2b0809788fbc/nav2_bt_navigator/behavior_trees/follow_waypoints_with_skip.xml) is an example
  * This is also an example of NavigateToPose BT-node
* ~~This PR contains refactoring of BtNavigator (to make the implementation of FollowWaypointsBtNavigator easier)~~ Reverted

## Use case
If there is an obstacle on the waypoint, the robot will not be able to reach it.
In such a case, I want to perform the waiting action used in [navigate_w_replanning_and_recovery.xml](https://github.com/ros-planning/navigation2/blob/9e304b1ad4c10d03f7a08991224fd9686c6bb26d/nav2_bt_navigator/behavior_trees/navigate_w_replanning_and_recovery.xml#L27).

## Description of documentation updates required from your changes

---

## Future work that may be required in bullet points

* Add unit and integration testing
* ~~Add test for plugins~~
* ~~Update nav2_behavior_tree/test/plugins/action/test_navigate_to_pose_action.cpp~~
* ~~Add documentation and BT diagrams~~
* ~~Add BT-nodes~~
* ~~Add sample BT (Use wait action)~~ Done.

### CODE REVIEW

File: nav2_behavior_tree/plugins/action/navigate_to_pose_action.cpp [hunk 1]
```diff
@@ -39,28 +39,15 @@ class NavigateToPoseAction : public BtActionNode<nav2_msgs::action::NavigateToPo
 
   void on_tick() override
   {
-    // Use the position and orientation fields from the XML attributes to initialize the goal
-    geometry_msgs::msg::Point position;
-    geometry_msgs::msg::Quaternion orientation;
-
-    if (!getInput("position", position) || !getInput("orientation", orientation)) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/navigate_to_pose_action.cpp:46]: Why did you remove error handling?

File: nav2_bt_navigator/CMakeLists.txt [hunk 2]
```diff
@@ -22,16 +22,23 @@ include_directories(
   include
 )
 
-set(executable_name bt_navigator)
+set(bt_navigator_library_name bt_navigator)
+set(navigate_to_pose_executable_name bt_navigator)
+set(follow_waypoints_executable_name follow_waypoints_bt_navigator)
```
Comment (SteveMacenski) [inline nav2_bt_navigator/CMakeLists.txt:27]: I don't understand why this is in this package, what is this for? If you're making a higher level application to do waypoint based following where the BT navigator is a node used in a larger behavior tree, this should be in another package to properly separate concerns.

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+
+<!--
+  BT to process FollowWaypoints.action
+-->
+<root main_tree_to_execute="MainTree">
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/follow_waypoints.xml:5]: This needs documentation added and a diagram of this tree to the readmes

File: nav2_bt_navigator/behavior_trees/follow_waypoints_with_skip.xml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+
+<!--
+  BT to process FollowWaypoints.action
+-->
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/follow_waypoints_with_skip.xml:4]: This needs documentation added and a diagram of this tree to the readmes

File: nav2_bt_navigator/include/nav2_bt_navigator/bt_navigator.hpp [hunk 5]
```diff
@@ -57,60 +52,36 @@ class BtNavigator : public nav2_util::LifecycleNode
    * @param state Reference to LifeCycle node state
    * @return SUCCESS or FAILURE
    */
-  nav2_util::CallbackReturn on_configure(const rclcpp_lifecycle::State & state) override;
+  virtual nav2_util::CallbackReturn on_configure(const rclcpp_lifecycle::State & state) override;
```
Comment (SteveMacenski) [inline nav2_bt_navigator/include/nav2_bt_navigator/bt_navigator.hpp:55]: Override means that its virtual, I'm pretty sure that this is going to fail linting

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > I don't understand why this is in this package, what is this for?

FollowWaypointsBtNavigator reads BT and Execute the FollowWaypoints action in the same way BtNavigator currently executes the NavigateToPose action.

> If you're making a higher level application

No, it's not application.  It only provides a framework for executing FollowWaypoints actions by BT.  (In addition, The use of NavigateToPose action node in BT is also user-dependent.)

File: nav2_behavior_tree/plugins/action/navigate_to_pose_action.cpp (hunk 1 above)
Comment (ymd-stella) [reply nav2_behavior_tree/plugins/action/navigate_to_pose_action.cpp:46]: Thanks for reviewing. I was careless.  
I've restored it.

File: nav2_bt_navigator/include/nav2_bt_navigator/bt_navigator.hpp (hunk 5 above)
Comment (ymd-stella) [reply nav2_bt_navigator/include/nav2_bt_navigator/bt_navigator.hpp]: Thanks. I've fixed it.

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > Should this potentially be in a new package

I think that both FollowWaypointsBtNavigator and NavigateToPoseBtNavigator does as well. What do you think?

**Comment by ymd-stella** [comment]
> I dont understand what that comment has to do with the above work.

BtActionNode has a symmetrical function with BtNavigator and is a hint of analogy about the function we need.

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt:27]: What do you mean both provided as plugins? For your follow case, you still need the normal BT navigator running under the hood for the `NavigateToPose` action server to follow the individual waypoints. Its not an either-or situation, this builds on top of that. 

"navigate" is how you get from A to B. A "navigator" is the thing responsible for getting you from A to B. If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > For your follow case, you still need the normal BT navigator running under the hood for the NavigateToPose action server to follow the individual waypoints.

Whether or not FollowWaypointsBtNavigator uses NavigateToPose is dependent on BT.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > "navigate" is how you get from A to B. A "navigator" is the thing responsible for getting you from A to B. If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

Thanks. If "navigate" is defined that way, then the structure and naming of this PR is wrong.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > What do you mean both provided as plugins?

I was considered to make the following structure.
- nav2_bt_navigator/bt_navigator.cpp/.hpp: Load a BT and provide interface to provide an action

plugins for bt_navigator
- nav2_bt_navigator/plugins/navigate_to_pose.cpp/.hpp: Load a BT to execute NavigateToPose  and provide an NavigateToPose 
- nav2_bt_navigator/plugins/follow_waypoints.cpp/.hpp: Load a BT to execute FollowWaypoints and provide an FollowWaypoints

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: > If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

- "Moving from A to B" is "navigate"
- "Moving from A to B while avoiding obstacles" is "navigate"
- "Moving from A to B via C" is "not navigate" ?
- "Moving from A to B via C (if there is an obstacle in C, move directly to B)" is "not navigate" ?

Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt:27]: A to B via C could be "navigate", yes, if its a via point or a constraint on the navigation request that you go through that point. However, I would not say that going to 3 distinct points A-B-C is "navigate" if they're distinct points for distinct reasons. If B though is a point needing to go through, without stopping, as a constraint on the request for some reason, then that would be 1 navigation request. This is all semantics though. 

What I'm trying to communicate is that I don't believe this should all be in 1 package and the naming is confusing. What you're doing is creating an application level behavior tree to implement a waypoint following task. There's absolutely no argument on my side that it is a useful demo application and some folks like yourself may choose to use it as their basic autonomy. My argument is how its being added to the project (or if it belongs in this project at all). 

My assertion is that this is a demo-autonomy feature similar to the `nav2_waypoint_follower`. It uses the behavior tree rather than an action server to call the individual navigation requests that make up its parts. I think it should be either in the nav2_waypoint_follower package, since it itself is a form of waypoint following, or as another new package doing `nav2_bt_waypoint_follower` or something. But I also do wonder if this belongs better in the `nav2_tutorials` or `nav2_demos` repos.

Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt:27]: @fmrico do you agree with my assessment here? You also make application-level behavior trees for your research. Does this fall in line with that or am I not seeing this in the right lens?

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:27]: I almost agree with you. I'll try to create `nav2_bt_waypoint_follower`.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: Done.

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      <SetCurrentGoal goals="{goals}" current_waypoint_idx="{current_waypoint_idx}" goal="{goal}" />
+      <NavigateToPose goal="{goal}" />
+      <!-- Return success if all goals are achieved, or update the `goal_updated` and `goal`, and then make it Running. -->
+      <ReactiveFallback name="CheckAndUpdateWaypointIndex">
+        <BlackboardCheckInt value_A="{current_waypoint_idx}" value_B="{max_waypoint_idx}" />
+        <Sequence>
+          <AddInt in_value_A="{current_waypoint_idx}" in_value_B="1" out_value="{current_waypoint_idx}" />
+          <KeepRunningUntilFailure>
+            <AlwaysSuccess />
+          </KeepRunningUntilFailure>
```
Comment (fmrico) [inline nav2_bt_navigator/behavior_trees/follow_waypoints.xml:18]: May I know the intention of this part?

File: nav2_bt_navigator/CMakeLists.txt [hunk 7]
```diff
@@ -22,16 +22,18 @@ include_directories(
...
+add_executable(${navigate_to_pose_executable_name}
+  src/navigate_to_pose/main.cpp
 )
 
-set(library_name ${executable_name}_core)
+set(library_name ${bt_navigator_library_name}_core)
 
 add_library(${library_name} SHARED
   src/bt_navigator.cpp
+  src/navigate_to_pose/bt_navigator.cpp
```
Comment (fmrico) [inline nav2_bt_navigator/CMakeLists.txt:36]: Are there two different files with the same name?

File: nav2_bt_navigator/behavior_trees/follow_waypoints_with_skip.xml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        <AddInt in_value_A="{current_waypoint_idx}" in_value_B="1" out_value="{current_waypoint_idx}" />
+      </RecoveryNode>
+      <!-- Return success if all goals are achieved, or update the `goal_updated` and `goal`, and then make it Running -->
+      <ReactiveFallback name="CheckAndUpdateWaypointIndex">
+        <BlackboardCheckInt value_A="{current_waypoint_idx}" value_B="{max_waypoint_idx}" />
+        <Sequence>
+          <AddInt in_value_A="{current_waypoint_idx}" in_value_B="1" out_value="{current_waypoint_idx}" />
+          <KeepRunningUntilFailure name="AlwaysRunning">
+            <AlwaysSuccess />
+          </KeepRunningUntilFailure>
```
Comment (fmrico) [inline nav2_bt_navigator/behavior_trees/follow_waypoints_with_skip.xml:24]: The same here

File: nav2_bt_navigator/behavior_trees/follow_waypoints_with_wait.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        <Wait wait_duration="5" />
+      </RecoveryNode>
+      <!-- Return success if all goals are achieved, or update the `goal_updated` and `goal`, and then make it Running -->
+      <ReactiveFallback name="CheckAndUpdateWaypointIndex">
+        <BlackboardCheckInt value_A="{current_waypoint_idx}" value_B="{max_waypoint_idx}" />
+        <Sequence>
+          <AddInt in_value_A="{current_waypoint_idx}" in_value_B="1" out_value="{current_waypoint_idx}" />
+          <KeepRunningUntilFailure name="AlwaysRunning">
+            <AlwaysSuccess />
+          </KeepRunningUntilFailure>
```
Comment (fmrico) [inline nav2_bt_navigator/behavior_trees/follow_waypoints_with_wait.xml:23]: The same here

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (fmrico) [reply nav2_bt_navigator/CMakeLists.txt]: @SteveMacenski, I agree with you. This PR should be at the application-level.

@ymd-stella is contributing to #1651 with an alternative to `nav2_waypoint_follower`, but using a BT. That's great. But it shouldn't affect core packages, like `bt_navigator.` This PR could be in tutorials or demos, in a separate package. If this demo requires changes in `bt_navigator`, maybe you could do it in different and concrete PRs, in which we could discuss its convenience, but not linked to a specific demo.

In a glance, I have many doubts in the XML files that @ymd-stella proposes. Many nodes are not implemented yet. I also consider that BT shouldn't contain nodes that do what it should be done in code. They must be of a higher level and simpler.

File: nav2_bt_navigator/CMakeLists.txt (hunk 7 above)
Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt:36]: - `nav2_bt_navigator/*/bt_navigator.cpp|.hpp`: Implementation of BtNavigator.
- `nav2_bt_navigator/*/navigate_to_pose/bt_navigator.cpp|.hpp`: Implementation of NavigateToPoseBtNavigator. NavigateToPoseBtNavigator inherits from BtNavigator.

I know it's a confusing name, so I'm thinking about it.

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml (hunk 6 above)
Comment (ymd-stella) [reply nav2_bt_navigator/behavior_trees/follow_waypoints.xml:18]: I want `AlwaysRunning`.
I think it's necessary to ensure that BT doesn't end when the goal is updated.

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: > If this demo requires changes in bt_navigator, maybe you could do it in different and concrete PRs, in which we could discuss its convenience, but not linked to a specific demo.

I'm thinking of changing nav2_bt_waypoint_follower to avoid depending on nav2_bt_navigator. If duplicate code is a problem for us, we will consider addressing it afterwards.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: I' ve removed the dependency from nav2_bt_waypoint_follower to nav2_bt_navigator.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: > Many nodes are not implemented yet.

Only two nodes.
- AddInt
- SetCurrentGoal

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml (hunk 6 above)
Comment (ymd-stella) [reply nav2_bt_navigator/behavior_trees/follow_waypoints.xml:18]: If you don't know about KeepRunningUntilFailure, please check https://github.com/BehaviorTree/BehaviorTree.CPP/pull/192.

Comment (ymd-stella) [reply nav2_bt_navigator/behavior_trees/follow_waypoints.xml:18]: I'm not familiar with BT design, so any advice would be appreciated.

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt]: I don't really mind how many nodes are proposed but not implemented, you're not going to be able to merge things into **production** that don't work.

File: nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_util/robot_utils.hpp"
+#include "nav2_behavior_tree/bt_conversions.hpp"
+#include "nav2_bt_waypoint_follower/ros_topic_logger.hpp"
+
+namespace nav2_bt_waypoint_follower
+{
+
+FollowWaypointsBtNavigator::FollowWaypointsBtNavigator()
+: nav2_util::LifecycleNode("bt_navigator", "", false)
```
Comment (SteveMacenski) [inline nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp:34]: This can't be named `bt_navigator`

File: nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+  // use suffix '_rclcpp_node' to keep parameter file consistency #1773
+  auto options = rclcpp::NodeOptions().arguments(
+    {"--ros-args",
+      "-r", std::string("__node:=") + get_name() + "_rclcpp_node",
+      "--"});
+  // Support for handling the topic-based goal pose from rviz
+  client_node_ = std::make_shared<rclcpp::Node>("_", options);
+
+  tf_ = std::make_shared<tf2_ros::Buffer>(get_clock());
```
Comment (SteveMacenski) [inline nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp:83]: Why do you need TF?

File: nav2_bt_waypoint_follower/src/ros_topic_logger.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <utility>
+#include "nav2_bt_waypoint_follower/ros_topic_logger.hpp"
+#include "tf2_ros/buffer_interface.h"
+
+namespace nav2_bt_navigator
```
Comment (SteveMacenski) [inline nav2_bt_waypoint_follower/src/ros_topic_logger.cpp:20]: This file looks like a complete copy paste, why not use the bt navigator version of this object?

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt]: You need to provide unit and integration testing for this.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: Of course I don't think I can merge this yet. I was just pointing this out because I felt that the implemented nodes were recognized differently.

Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: > You need to provide unit and integration testing for this.

I'll try it.

File: nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp (hunk 10 above)
Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp]: Fixed it in 0d950ea.

File: nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp (hunk 11 above)
Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/bt_waypoint_follower.cpp]: Removed it.

File: nav2_bt_waypoint_follower/src/ros_topic_logger.cpp (hunk 12 above)
Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/ros_topic_logger.cpp:20]: bt_navigator is not exporting this.
Based on https://github.com/ros-planning/navigation2/pull/1796#discussion_r439714103,
> If this demo requires changes in bt_navigator, maybe you could do it in different and concrete PRs, in which we could discuss its convenience, but not linked to a specific demo.

it is needed to avoid making changes to bt_navigator.

Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/ros_topic_logger.cpp:20]: If it's ok to change bt_navigator, I'll change `bt_navigator/CMakeLists.txt` and reuse the code.

Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/ros_topic_logger.cpp:20]: Or should I create another PR?

**Comment by ymd-stella** [comment]
@fmrico thanks for reviewing.
> I also consider that BT shouldn't contain nodes that do what it should be done in code. They must be of a higher level and simpler.

Is this better?
```
<root main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <ReactiveFallback name="FollowWaypoints">
      <AllGoalAchieved />
      <KeepRunningUntilFailure>
        <Sequence>
          <GetNextGoal goals="{goals}" goal="{goal}" />
          <NavigateToPose goal="{goal}" />
        </Sequence>
      </KeepRunningUntilFailure>
    </ReactiveFallback>
  </BehaviorTree>
</root>
```

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (ymd-stella) [reply nav2_bt_navigator/CMakeLists.txt]: Resolved this conversation?

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml (hunk 6 above)
Comment (fmrico) [reply nav2_bt_navigator/behavior_trees/follow_waypoints.xml]: I implemented `KeepRunningUntilFailure` :)

My comment is for avoiding hacks in BTs.

File: nav2_bt_navigator/CMakeLists.txt (hunk 2 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/CMakeLists.txt]: Got it, I was adding it here since we were having a thematic conversation here, we can move to the main thread now

**Comment by SteveMacenski** [comment]
I wonder if we should actually rename this to be slightly more general. While the BTs provided are specific to waypoint following, I think this could be the basis for an autonomy behavior tree application, where one such application is the waypoint follower. E.g. we could also have this same package run a BT that asks a robot to go from A->B->moveit to pick up box->C->moveit to drop off box. This could be the generic application server. @ymd-stella is there anything specific here (other than the action server `FollowWaypoints` which could be a parameter) that makes this waypoint follower specific other than the behavior tree? 

Having an example application that could be built on top of for autonomy applications would be something unique to ROS2 navigation @fmrico what do you think? Is there enough boiler plate that this makes sense to do and do you think that a generic application-BT-hosting-server package makes sense to include?

**Comment by ymd-stella** [comment]
> Is there anything specific here (other than the action server FollowWaypoints which could be a parameter) that makes this waypoint follower specific other than the behavior tree?

- How to get feedback (current_waypoint in FollowWaypoints, current_pose/navigation_time/number_of_recoveries/distance_remaining in NavigateToPose).
- How to initialize blackboard.

> Maybe if there are some specifics we can make a BTApplications templated/base class that the waypoint follower or move-mover application implement/derive from. I'm thinking maybe the action server template would make us do one of those.

They had different names BTApplications/BtNavigatorBase, but before d973c4e1 (https://github.com/ros-planning/navigation2/compare/master...d973c4e1 ), I was tried to do it. I should have created a PR that generalizes BtNavigator first, and then considered creating a PR for FollowWaypoints.

**Comment by SteveMacenski** [comment]
Those could all be implemented through an abstract interface using the `on_XYZ()` pattern. 

I'm not sure that I agree that BTNavigator and this work needs to share much code / base. Their roles are very different and the only thing they share is setting up and using behavior trees. The action servers serve different purposes, this probably doesn't need TF or goal subscriptions, etc. We could move some of the loading and parsing of BTs to the `nav2_behavior_tree` library for both to use, but other than that, I think we should be thinking of these as completely separate ideas.

I think we should instead focus on the generalization of a behavior tree application framework, where your waypoint follower is a key demo of that. The BT Action Node I think is a good example of the pattern that we'd need to implement this. A template based on the `ActionT` and then virtual `on_XYZ()` functions to setup the `ActionT` messages and implement the behavior-specific items. That would also give designers a clear state machine architecture to work in with their more complex behavior trees.

File: nav2_bt_navigator/behavior_trees/follow_waypoints.xml (hunk 3 above)
Comment (ymd-stella) [reply nav2_bt_navigator/behavior_trees/follow_waypoints.xml]: Added to `Future work` of description.

File: nav2_bt_waypoint_follower/src/ros_topic_logger.cpp (hunk 12 above)
Comment (ymd-stella) [reply nav2_bt_waypoint_follower/src/ros_topic_logger.cpp]: Removed codes that was using RosTopicLogger. We'll add it if we ever need it.

**Comment by SteveMacenski** [comment]
Closing Pr, we’ve had no updates in awhile and it seems like this is your development branch.

**Comment by ymd-stella** [comment]
No problem.
I'll create the PR again when I add the test.
Thanks.
